### PR TITLE
Api support

### PR DESCRIPTION
**Pull request by ti129**
1. SPRAVA api support
2. Loading animation
3. Error handling api request
4. Caching API json and imgs.

### CODE REVIEW

File: src/actions/index.js [hunk 1]
```diff
@@ -7,3 +9,46 @@ export const VisibilityFilters = {
   SHOW_ALL: 'SHOW_ALL',
   SHOW_SAVED: 'SHOW_SAVED',
 };
+
+// action creators
+
+export function itemsHasErrored(bool) {
```
Comment (davojta) [inline src/actions/index.js]: you can take a look at redux-ducks and for FSA standart for the action and later refactor the code to them

File: src/components/checkbox/checkbox.scss [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 .check__box {
   position: absolute;
-  margin-top: 0.2em;
+  margin-top: -3.8em;
```
Comment (davojta) [inline src/components/checkbox/checkbox.scss]: it is too later for rem, what do you think ?

File: src/components/eventList/eventList.js [hunk 3]
```diff
@@ -1,34 +1,100 @@
...
-        description={e.description}
-        isEventSave={Boolean(e.checked)}
-      />
-    ))}
-  </div>
-);
+class EventList extends React.Component {
+  componentDidMount() {
+    const { fetchData } = this.props;
+    fetchData('https://cors-anywhere.herokuapp.com/http://spravafestival.by/api/music');
```
Comment (davojta) [inline src/components/eventList/eventList.js]: the url to const or config ?

File: src/components/eventList/eventList.js [hunk 4]
```diff
@@ -1,34 +1,100 @@
...
+    if (isLoading) {
+      return <Loading />;
+    }
+
+    return (
+      <div className="events-section__event-list-container" id="event-list">
+        <div
+          className="events-section__first-day-events"
+          style={(Object.keys(events)
+            .filter(e => events[e].date === '4 АВГУСТА').length > 0)
```
Comment (davojta) [inline src/components/eventList/eventList.js]: 4 of august to config ?

**Review by davojta — APPROVED** [review]
it is okay to me to merge the PR right now and later to refactor the code

File: src/actions/index.js (hunk 1 above)
Comment (ti129) [reply src/actions/index.js]: I should follow one of thise patterns to write the actions.

Comment (davojta) [reply src/actions/index.js]: I'm sure that it worth to follow both patterns )

Comment (ti129) [reply src/actions/index.js]: If I have more than one action in the file. In this app I have three actions for three different property of app state, there are: `itemsHasErrored`, `itemsIsLoading`, `itemsFetchDataSuccess`. And all these actions work with one thing, with API. And all actions located in one file, and all reducers is located in one file. (Other reducers located in other files).

If I'll follow the ducks pattern I should to located all reducers in different files. And all actions shold be located in different files. And I think that it is too much.

I suggest this:
image

**Comment by davojta** [comment]
Seems you should have 1 duck for data + 3 actions in one file )

**Comment by ti129** [comment]
Refactor is done

**Comment by ti129** [comment]
And one problem with images. My antivirus firewall blocks request to pages without htttp. And the site SPRAVA without htttp. And now images can't load.

File: src/actions/index.js (hunk 1 above)
Comment (davojta) [reply src/actions/index.js]: @AlexandrTichon Why the api reducer cant be duck? We can call each other and discuss the question by voice )

File: src/actions/apiFetch.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+export const itemsStatus = (itemsStatusData, dataSuccess = {}) => ({
+  type: ACTION_TYPES.SET_ITEMS_STATUS,
+  payload: {
+    itemsStatus: itemsStatusData,
+    items: dataSuccess,
+  },
+});
+
+export const itemsFetchData = url => (dispatch) => {
```
Comment (davojta) [inline src/actions/apiFetch.js]: url can be just a const in the file on top level

```
const URL_TO_FETCH_STRAVA_DATA = '';
export const itemsFetchData = url => (dispatch) => {
```

File: src/actions/apiFetch.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import 'whatwg-fetch';
```
Comment (davojta) [inline src/actions/apiFetch.js:1]: why do we need that ?

File: src/actions/apiFetch.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    ITEMS_HAS_ERRORED,
+    ITEMS_IS_LOADING,
+    ITEMS_FETCH_DATA_SUCCESS,
+  } = ITEMS_STATUS_PARAMS;
+  dispatch(itemsStatus(ITEMS_IS_LOADING));
+
+  window.fetch(url, { cache: 'force-cache' })
+    .then((response) => {
+      if (!response.ok) {
+        throw Error(response.statusText);
```
Comment (davojta) [inline src/actions/apiFetch.js:31]: need add console.error with text readable for human 
```
console.error('some error with response in fetch data from server')
```

File: src/components/eventList/eventList.js [hunk 8]
```diff
@@ -1,34 +1,101 @@
...
+    if (fetchStatus === ITEMS_STATUS_PARAMS.ITEMS_IS_LOADING) {
+      return <Loading />;
+    }
+
+    return (
+      <div className="events-section__event-list-container" id="event-list">
+        <div
+          className="events-section__first-day-events"
+          style={(Object.keys(events)
+            .filter(e => events[e].date === `${FESTIVAL_DAYS.FIRST_DAY}`).length > 0)
```
Comment (davojta) [inline src/components/eventList/eventList.js]: could we move that filter to separate function ? it is too long and hard to read among the markup

File: src/constants/constants.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+export const SPRAVA_API_MAIN_URL = 'http://spravafestival.by/api';
+
+export const CORS_URL = 'https://cors-anywhere.herokuapp.com';
```
Comment (davojta) [inline src/constants/constants.js:3]: +1

File: src/actions/apiFetch.js (hunk 6 above)
Comment (ti129) [reply src/actions/apiFetch.js:1]: It's fetch polyfill

Comment (davojta) [reply src/actions/apiFetch.js:1]: I mean why we need polyfill here ? ) i think it is okey to support only evergreen browsers and do not support ie11 )

**Pull request merged**